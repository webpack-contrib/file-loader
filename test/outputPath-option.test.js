import {
  compile,
  execute,
  getCompiler,
  normalizeErrors,
  readAsset,
} from './helpers';

describe('"outputPath" option', () => {
  it('should work without value', async () => {
    const compiler = getCompiler('simple.js');
    const stats = await compile(compiler);

    expect(
      execute(readAsset('main.bundle.js', compiler, stats))
    ).toMatchSnapshot('result');
    expect(Object.keys(stats.compilation.assets)).toMatchSnapshot('assets');
    expect(normalizeErrors(stats.compilation.warnings)).toMatchSnapshot(
      'warnings'
    );
    expect(normalizeErrors(stats.compilation.errors)).toMatchSnapshot('errors');
  });

  it('should work with "String" value', async () => {
    const compiler = getCompiler('simple.js', {
      outputPath: 'output_path/',
    });
    const stats = await compile(compiler);

    expect(
      execute(readAsset('main.bundle.js', compiler, stats))
    ).toMatchSnapshot('result');
    expect(Object.keys(stats.compilation.assets)).toMatchSnapshot('assets');
    expect(normalizeErrors(stats.compilation.warnings)).toMatchSnapshot(
      'warnings'
    );
    expect(normalizeErrors(stats.compilation.errors)).toMatchSnapshot('errors');
  });

  it('should work with "String" value without trailing slash', async () => {
    const compiler = getCompiler('simple.js', {
      outputPath: 'output_path',
    });
    const stats = await compile(compiler);

    expect(
      execute(readAsset('main.bundle.js', compiler, stats))
    ).toMatchSnapshot('result');
    expect(Object.keys(stats.compilation.assets)).toMatchSnapshot('assets');
    expect(normalizeErrors(stats.compilation.warnings)).toMatchSnapshot(
      'warnings'
    );
    expect(normalizeErrors(stats.compilation.errors)).toMatchSnapshot('errors');
  });

  it('should work with "String" value and with "name" option', async () => {
    const compiler = getCompiler('simple.js', {
      name: '[path][name].[ext]',
      outputPath: 'output_path/',
    });
    const stats = await compile(compiler);

    expect(
      execute(readAsset('main.bundle.js', compiler, stats))
    ).toMatchSnapshot('result');
    expect(Object.keys(stats.compilation.assets)).toMatchSnapshot('assets');
    expect(normalizeErrors(stats.compilation.warnings)).toMatchSnapshot(
      'warnings'
    );
    expect(normalizeErrors(stats.compilation.errors)).toMatchSnapshot('errors');
  });

  it('should work with "Function" value', async () => {
    // TODO
    const compiler = getCompiler('simple.js', {
      outputPath(url, resourcePath, context) {
        expect(url).toBe('9c87cbf3ba33126ffd25ae7f2f6bbafb.png');
        expect(resourcePath).toMatch('file.png');
        expect(context).toMatch('fixtures');

        return `output_path_func/${url}`;
      },
    });
    const stats = await compile(compiler);

    expect(
      execute(readAsset('main.bundle.js', compiler, stats))
    ).toMatchSnapshot('result');
    expect(Object.keys(stats.compilation.assets)).toMatchSnapshot('assets');
    expect(normalizeErrors(stats.compilation.warnings)).toMatchSnapshot(
      'warnings'
    );
    expect(normalizeErrors(stats.compilation.errors)).toMatchSnapshot('errors');
  });

  it('should work with "Function" value and with "name" option', async () => {
    const compiler = getCompiler('simple.js', {
      name: '[name].[ext]',
      outputPath(url) {
        return `output_path_func/${url}`;
      },
    });
    const stats = await compile(compiler);

    expect(
      execute(readAsset('main.bundle.js', compiler, stats))
    ).toMatchSnapshot('result');
    expect(Object.keys(stats.compilation.assets)).toMatchSnapshot('assets');
    expect(normalizeErrors(stats.compilation.warnings)).toMatchSnapshot(
      'warnings'
    );
    expect(normalizeErrors(stats.compilation.errors)).toMatchSnapshot('errors');
  });

  it('should work with "String" value and with "publicPath" option', async () => {
    const compiler = getCompiler('simple.js', {
      outputPath: 'output_path/',
      publicPath: 'public_path/',
    });
    const stats = await compile(compiler);

    expect(
      execute(readAsset('main.bundle.js', compiler, stats))
    ).toMatchSnapshot('result');
    expect(Object.keys(stats.compilation.assets)).toMatchSnapshot('assets');
    expect(normalizeErrors(stats.compilation.warnings)).toMatchSnapshot(
      'warnings'
    );
    expect(normalizeErrors(stats.compilation.errors)).toMatchSnapshot('errors');
  });

  it('should work with "String" value and with "publicPath" option and without trailing slash', async () => {
    const compiler = getCompiler('simple.js', {
      outputPath: 'output_path',
      publicPath: 'public_path',
    });
    const stats = await compile(compiler);

    expect(
      execute(readAsset('main.bundle.js', compiler, stats))
    ).toMatchSnapshot('result');
    expect(Object.keys(stats.compilation.assets)).toMatchSnapshot('assets');
    expect(normalizeErrors(stats.compilation.warnings)).toMatchSnapshot(
      'warnings'
    );
    expect(normalizeErrors(stats.compilation.errors)).toMatchSnapshot('errors');
  });

  it('should work with "Function" value and with "publicPath" as "String"', async () => {
    const compiler = getCompiler('simple.js', {
      publicPath: 'public_path/',
      outputPath(url) {
        return `output_path_func/${url}`;
      },
    });
    const stats = await compile(compiler);

    expect(
      execute(readAsset('main.bundle.js', compiler, stats))
    ).toMatchSnapshot('result');
    expect(Object.keys(stats.compilation.assets)).toMatchSnapshot('assets');
    expect(normalizeErrors(stats.compilation.warnings)).toMatchSnapshot(
      'warnings'
    );
    expect(normalizeErrors(stats.compilation.errors)).toMatchSnapshot('errors');
  });

  it('should work with "String" value and with "publicPath" as "Function"', async () => {
    const compiler = getCompiler('simple.js', {
      publicPath(url) {
        return `public_path_func/${url}`;
      },
      outputPath: 'output_path/',
    });
    const stats = await compile(compiler);

    expect(
      execute(readAsset('main.bundle.js', compiler, stats))
    ).toMatchSnapshot('result');
    expect(Object.keys(stats.compilation.assets)).toMatchSnapshot('assets');
    expect(normalizeErrors(stats.compilation.warnings)).toMatchSnapshot(
      'warnings'
    );
    expect(normalizeErrors(stats.compilation.errors)).toMatchSnapshot('errors');
  });

  it('should work with "Function" value and with "publicPath" as "Function"', async () => {
    const compiler = getCompiler('simple.js', {
      outputPath(url) {
        return `output_path_func/${url}`;
      },
      publicPath(url) {
        return `public_path_func/${url}`;
      },
    });
    const stats = await compile(compiler);

    expect(
      execute(readAsset('main.bundle.js', compiler, stats))
    ).toMatchSnapshot('result');
    expect(Object.keys(stats.compilation.assets)).toMatchSnapshot('assets');
    expect(normalizeErrors(stats.compilation.warnings)).toMatchSnapshot(
      'warnings'
    );
    expect(normalizeErrors(stats.compilation.errors)).toMatchSnapshot('errors');
  });
});
