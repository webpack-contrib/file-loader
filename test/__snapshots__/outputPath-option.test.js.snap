// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`"outputPath" option should work with "Function" value and with "name" option: assets 1`] = `
Array [
  "output_path_func/file.png",
  "main.bundle.js",
]
`;

exports[`"outputPath" option should work with "Function" value and with "name" option: errors 1`] = `Array []`;

exports[`"outputPath" option should work with "Function" value and with "name" option: result 1`] = `"output_path_func/file.png"`;

exports[`"outputPath" option should work with "Function" value and with "name" option: warnings 1`] = `Array []`;

exports[`"outputPath" option should work with "Function" value and with "publicPath" as "Function": assets 1`] = `
Array [
  "output_path_func/39f5c21c1aee6ff21844c6e1d8251d97.png",
  "main.bundle.js",
]
`;

exports[`"outputPath" option should work with "Function" value and with "publicPath" as "Function": errors 1`] = `Array []`;

exports[`"outputPath" option should work with "Function" value and with "publicPath" as "Function": result 1`] = `"public_path_func/39f5c21c1aee6ff21844c6e1d8251d97.png"`;

exports[`"outputPath" option should work with "Function" value and with "publicPath" as "Function": warnings 1`] = `Array []`;

exports[`"outputPath" option should work with "Function" value and with "publicPath" as "String": assets 1`] = `
Array [
  "output_path_func/39f5c21c1aee6ff21844c6e1d8251d97.png",
  "main.bundle.js",
]
`;

exports[`"outputPath" option should work with "Function" value and with "publicPath" as "String": errors 1`] = `Array []`;

exports[`"outputPath" option should work with "Function" value and with "publicPath" as "String": result 1`] = `"public_path/39f5c21c1aee6ff21844c6e1d8251d97.png"`;

exports[`"outputPath" option should work with "Function" value and with "publicPath" as "String": warnings 1`] = `Array []`;

exports[`"outputPath" option should work with "Function" value: assets 1`] = `
Array [
  "output_path_func/39f5c21c1aee6ff21844c6e1d8251d97.png",
  "main.bundle.js",
]
`;

exports[`"outputPath" option should work with "Function" value: errors 1`] = `Array []`;

exports[`"outputPath" option should work with "Function" value: result 1`] = `"output_path_func/39f5c21c1aee6ff21844c6e1d8251d97.png"`;

exports[`"outputPath" option should work with "Function" value: warnings 1`] = `Array []`;

exports[`"outputPath" option should work with "String" value and with "name" option: assets 1`] = `
Array [
  "output_path/file.png",
  "main.bundle.js",
]
`;

exports[`"outputPath" option should work with "String" value and with "name" option: errors 1`] = `Array []`;

exports[`"outputPath" option should work with "String" value and with "name" option: result 1`] = `"output_path/file.png"`;

exports[`"outputPath" option should work with "String" value and with "name" option: warnings 1`] = `Array []`;

exports[`"outputPath" option should work with "String" value and with "publicPath" as "Function": assets 1`] = `
Array [
  "output_path/39f5c21c1aee6ff21844c6e1d8251d97.png",
  "main.bundle.js",
]
`;

exports[`"outputPath" option should work with "String" value and with "publicPath" as "Function": errors 1`] = `Array []`;

exports[`"outputPath" option should work with "String" value and with "publicPath" as "Function": result 1`] = `"public_path_func/39f5c21c1aee6ff21844c6e1d8251d97.png"`;

exports[`"outputPath" option should work with "String" value and with "publicPath" as "Function": warnings 1`] = `Array []`;

exports[`"outputPath" option should work with "String" value and with "publicPath" option and without trailing slash: assets 1`] = `
Array [
  "output_path/39f5c21c1aee6ff21844c6e1d8251d97.png",
  "main.bundle.js",
]
`;

exports[`"outputPath" option should work with "String" value and with "publicPath" option and without trailing slash: errors 1`] = `Array []`;

exports[`"outputPath" option should work with "String" value and with "publicPath" option and without trailing slash: result 1`] = `"public_path/39f5c21c1aee6ff21844c6e1d8251d97.png"`;

exports[`"outputPath" option should work with "String" value and with "publicPath" option and without trailing slash: warnings 1`] = `Array []`;

exports[`"outputPath" option should work with "String" value and with "publicPath" option: assets 1`] = `
Array [
  "output_path/39f5c21c1aee6ff21844c6e1d8251d97.png",
  "main.bundle.js",
]
`;

exports[`"outputPath" option should work with "String" value and with "publicPath" option: errors 1`] = `Array []`;

exports[`"outputPath" option should work with "String" value and with "publicPath" option: result 1`] = `"public_path/39f5c21c1aee6ff21844c6e1d8251d97.png"`;

exports[`"outputPath" option should work with "String" value and with "publicPath" option: warnings 1`] = `Array []`;

exports[`"outputPath" option should work with "String" value without trailing slash: assets 1`] = `
Array [
  "output_path/39f5c21c1aee6ff21844c6e1d8251d97.png",
  "main.bundle.js",
]
`;

exports[`"outputPath" option should work with "String" value without trailing slash: errors 1`] = `Array []`;

exports[`"outputPath" option should work with "String" value without trailing slash: result 1`] = `"output_path/39f5c21c1aee6ff21844c6e1d8251d97.png"`;

exports[`"outputPath" option should work with "String" value without trailing slash: warnings 1`] = `Array []`;

exports[`"outputPath" option should work with "String" value: assets 1`] = `
Array [
  "output_path/39f5c21c1aee6ff21844c6e1d8251d97.png",
  "main.bundle.js",
]
`;

exports[`"outputPath" option should work with "String" value: errors 1`] = `Array []`;

exports[`"outputPath" option should work with "String" value: result 1`] = `"output_path/39f5c21c1aee6ff21844c6e1d8251d97.png"`;

exports[`"outputPath" option should work with "String" value: warnings 1`] = `Array []`;

exports[`"outputPath" option should work without value: assets 1`] = `
Array [
  "39f5c21c1aee6ff21844c6e1d8251d97.png",
  "main.bundle.js",
]
`;

exports[`"outputPath" option should work without value: errors 1`] = `Array []`;

exports[`"outputPath" option should work without value: result 1`] = `"39f5c21c1aee6ff21844c6e1d8251d97.png"`;

exports[`"outputPath" option should work without value: warnings 1`] = `Array []`;
